Modules/arraymodule.c

```diff
@@ -32,18 +34,22 @@ static struct PyModuleDef arraymodule;
 struct arraydescr {
     char typecode;
     int itemsize;
-    PyObject * (*getitem)(struct arrayobject *, Py_ssize_t);
-    int (*setitem)(struct arrayobject *, Py_ssize_t, PyObject *);
+    PyObject * (*getitem)(char *, Py_ssize_t);
+    int (*setitem)(char *, Py_ssize_t, PyObject *);
     int (*compareitems)(const void *, const void *, Py_ssize_t);
     const char *formats;
     int is_integer_type;
     int is_signed;
 };
 
+typedef struct {
+    Py_ssize_t allocated;
+    char items[];
```

Comment:
I think this has potential alignment issues that didn't exist previously. For example, on 32-bit platforms `allocated` will be four bytes. If the array stores doubles, they will be aligned to 4 bytes instead of the required 8 bytes.

I think we need an `_Alignas()` for items. Something like:

```
typedef struct {
    Py_ssize_t allocated;
    _Alignas(8) char items[];
} arraydata;
```